bindings/rust/s2n-tls/src/session_ticket.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use core::ptr::NonNull;
+
+use s2n_tls_sys::*;
+
+use crate::{
+    connection::Connection,
+    error::{Error, Fallible},
+};
+
+/// A trait to retrieve session tickets from the connection
+pub trait SessionTicketCallback {
+    fn on_session_ticket(&self, connection: &mut Connection, session_ticket: SessionTicket);
+}
+
+/// Wrapper around a session ticket struct
+#[derive(Debug)]
+pub struct SessionTicket {
+    ticket: NonNull<s2n_session_ticket>,
+}
```

Comment:
Instead of wrapping a `s2n_session_ticket` pointer, could we reuse the trick from the ClientHello bindings where we directly alias the type?
https://github.com/aws/s2n-tls/blob/65e74ca7c116ca40f81732c0906f6419e8c7bfa5/bindings/rust/s2n-tls/src/client_hello.rs#L55

I think this generally makes some of our callback things safer, since we then would be returning reference to the session ticket rather than ownership of the session ticket.

This mostly impacts the `SessionTicketCallback`, where `on_session_ticket` would have a reference to a session_ticket rather than the owned session ticket. This would prevent clients from doing unsafe things like storing the session ticket pointer in persistent memory like a vector or something.

Reply:
Actually, see my comment above. My new preference is for the Rust bindings to have a totally separate struct that actually _is_ owned by the application. This does make our bindings a little bit thicker and isn't our traditional way of doing things, but I still think it's worth considering.

Reply:
Or better yet, we could mostly remove the handling of session tickets from client concerns. I really like the way rustls does this: https://docs.rs/rustls/latest/rustls/client/trait.ClientSessionStore.html

Reply:
Having trouble following this thread... It sounds like you no longer want the callback to provide a reference to the user. I'm going to resolve this comment so that I can narrow down what your suggestion is; its just sort of confusing to have two comments recommend the opposite thing.